File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        L, R = 0, len(nums) - 1
        res = 1001
        while L <= R:
            M = (L + R) // 2
            res = min(res, nums[M])

            if nums[L] < nums[R]:
                res = min(res, nums[L])
            
            if nums[L] <= nums[M]:
                L = M + 1
            elif nums[M] < nums[R]:
                R = M - 1

        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // [1, 2, 4, 6]
        // [1, 1, 1, 1]
        // [1, 1, 2, 8]
        // [48, 24, 6, 1]
        // [48, 24, 12, 8]
        vector<int> res(nums.size(), 1);
        int post = 1;
        for(int i = res.size()-1; i >= 0; i--){
            res[i] = post;
            post *= nums[i];
        }
        int pre = 1;
        for(int i = 0; i < res.size(); i++){
            res[i] *= pre;
            pre *= nums[i];
        }

        return res;
    }
};
